File: frontend/src/types/finances/transaction.ts
export interface Transaction {
    id: number;
    destination: string;
    amount: number;
    cashback: number;
    replenishment: boolean;
    transaction_date: string;
    category_id: number;
    account_id: number;
    user_id: number;
}

export interface TransactionCreate {
    destination: string;
    amount: number;
    cashback: number;
    replenishment: boolean;
    transaction_date: string;
    category_id: number;
    account_id: number;
}

export interface TransactionChange {
    destination?: string;
    amount?: number;
    cashback?: number;
    replenishment?: boolean;
    transaction_date?: string;
    category_id?: number;
    account_id?: number;
}

export interface TransactionFilter {
    page?: number;
    size?: number;
    destination?: string;
    min_amount?: number;
    max_amount?: number;
    amount?: number;
    min_cashback?: number;
    max_cashback?: number;
    cashback?: number;
    transaction_date_from?: string;
    transaction_date_to?: string;
    transaction_date?: string;
    category_id?: number[];
    account_id?: number[];
}

File: backend/finance_app/app/schemas/finances/transaction.py
from datetime import date
from decimal import Decimal
from fastapi import Query
from typing import Optional, List, Annotated

from pydantic import BaseModel, ConfigDict


class TransactionCreate(BaseModel):
    destination: str
    amount: Decimal
    cashback: Decimal
    replenishment: bool
    transaction_date: date
    category_id: int
    account_id: int


class TransactionChange(BaseModel):
    destination: Optional[str] = None
    amount: Optional[Decimal] = None
    cashback: Optional[Decimal] = None
    replenishment: Optional[bool] = None
    transaction_date: Optional[date] = None
    category_id: Optional[int] = None
    account_id: Optional[int] = None


class TransactionRead(BaseModel):
    id: int
    destination: str
    amount: Decimal
    cashback: Decimal
    replenishment: bool
    transaction_date: date
    category_id: int
    account_id: int
    user_id: int

    model_config = ConfigDict(from_attributes=True)


class TransactionFilter(BaseModel):
    destination: Optional[str] = None
    min_amount: Optional[Decimal] = None
    max_amount: Optional[Decimal] = None
    amount: Optional[Decimal] = None
    min_cashback: Optional[Decimal] = None
    max_cashback: Optional[Decimal] = None
    cashback: Optional[Decimal] = None
    transaction_date_from: Optional[date] = None
    transaction_date_to: Optional[date] = None
    transaction_date: Optional[date] = None
    category_id: Annotated[Optional[List[int]], Query()] = None
    account_id: Annotated[Optional[List[int]], Query()] = None
